cdap-api/src/main/java/io/cdap/cdap/api/app/ApplicationUpdateContext.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright © 2020 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.api.app;
+
+import com.google.gson.JsonSyntaxException;
+import io.cdap.cdap.api.Config;
+import io.cdap.cdap.api.artifact.ArtifactId;
+
+import io.cdap.cdap.api.artifact.ArtifactScope;
+import io.cdap.cdap.api.artifact.ArtifactVersionRange;
+import java.lang.reflect.Type;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * Context for updating Application configs. Provides helper methods for application to support operations like config
+ * upgrade.
+ */
+public interface ApplicationUpdateContext {
+
+  /**
+   * @return All update actions application should perform on the config.
+   */
+  List<ApplicationConfigUpdateAction> getConfigUpdateActions();
+
+  /**
+   * Get the old config as an object of the given type. The platform would perform the json deserialization based on
+   * the provided type. This is for the case where an application has the same/compatible/old config class. Application
+   * should decide on how they want to convert config from old to current type.
+   *
+   * @param configType type of the config platform should deserialize to.
+   * @return application config serialized to an object of given configType.
+   */
+  <C extends Config> C getConfig(Type configType) throws JsonSyntaxException;
```

Comment:
should not throw a gson exception, we want to keep the api module as clean as possible. Should throw some common java exception, like IllegalArgumentException.

In general, don't need to put runtime exceptions in the method signature. Can just document in the javadoc with `@throws` what type of runtime exceptions can be thrown.

Reply:
Think it makes sense to add a getOldArtifact() method that returns the app artifact that we are upgrading from.

Otherwise, the app may not know what class to deserialize this into. Taking data pipeline as an example, it would deserialize configs from 3.x differently than 4.x, etc.

Not saying we need to support pipeline upgrade from those old versions, but we may have similar situations in the future for newer apps like CDC.

Reply:
since it's now update, maybe 'oldArtifact' is a bad name

Reply:
Hey Albert, I will work with you on adding this new API you mentioned. Submitting this PR for now to get started on the follow up work.